core/src/main/java/com/linecorp/armeria/common/stream/AbstractStreamMessageDuplicator.java

```diff
@@ -902,10 +920,15 @@ private void doubleCapacity() {
             this.elements = a;
         }
 
-        // Steals a reference to the object.
+        /**
+         * Invoked by the executor in {@link DownstreamSubscription}.
+         */
         Object get(int offset) {
             final int head = this.head;
             final int tail = this.tail;
+            final Object[] elements = this.elements;
+            assert elements != null : "elements is null. SignalQueue: " + this;
+
             final int length = elements.length;
```

Comment:
This was the problem. The reference to this `elements` can be changed before the current thread reaches line 943.